nemo/collections/asr/data/huggingface/hf_audio_to_text.py

```diff
@@ -0,0 +1,592 @@
+            sample_rate=sample_rate,
+            augmentor=augmentor,
+            trim=trim,
+            bos_id=bos_id,
+            eos_id=eos_id,
+            pad_id=pad_id,
+            return_sample_id=return_sample_id,
+            channel_selector=channel_selector,
+            normalize_db=normalize_db,
+            ref_channel=ref_channel,
+            id_key=id_key,
+            normalize_text=normalize_text,
+            symbols_to_keep=symbols_to_keep,
+        )
+
+
+class _HFIterableAudioTextDataset(IterableDataset):
+    def __init__(
+        self,
+        audio_key: str,
+        text_key: str,
+        sample_rate_key: str,
+        hf_data_cfg: DictConfig,
+        parser: Union[str, Callable],
+        sample_rate: int,
+        augmentor: 'nemo.collections.asr.parts.perturb.AudioAugmentor' = None,
+        trim: bool = False,
+        bos_id: Optional[int] = None,
+        eos_id: Optional[int] = None,
+        pad_id: int = 0,
+        return_sample_id: bool = False,
+        id_key: Optional[str] = None,
+        channel_selector: Optional[ChannelSelectorType] = None,
+        normalize_db: Optional[float] = None,
+        ref_channel: Optional[int] = None,
+        global_rank: int = 0,
+        world_size: int = 0,
+        shuffle_n: int = 0,
+        shuffle_seed: int = 1234,
+        normalize_text: bool = False,
+        symbols_to_keep: Optional[str] = None,
+    ) -> None:
+        super().__init__()
+
+        if return_sample_id and id_key is None:
+            raise ValueError("return_sample_id is True, but id_key is None")
+
+        self.audio_key = audio_key
+        self.text_key = text_key
+        self.sample_rate_key = sample_rate_key
+        self.id_key = id_key
+        self.sample_rate = sample_rate
+        self.augmentor = augmentor if augmentor is not None else AudioAugmentor()
+        self.trim = trim
+        self.return_sample_id = return_sample_id
+        self.channel_selector = channel_selector
+        self.normalize_db = normalize_db
+        self.ref_channel = ref_channel
+
+        self.text_processor = HFTextProcessor(parser, bos_id, eos_id, pad_id, normalize_text, symbols_to_keep)
+
+        with open_dict(hf_data_cfg):
+            # streaming must be True for iterable dataset
+            hf_data_cfg.streaming = True
+
+        logging.info(f"Using HuggingFace IterableDataset with cfg: {hf_data_cfg}")
+        self.dataset = hf_datasets.load_dataset(**hf_data_cfg)
+
+        if shuffle_n > 0:
+            self.dataset = self.dataset.shuffle(seed=shuffle_seed, buffer_size=shuffle_n)
+
+        self.dataset = split_dataset_by_node(self.dataset, global_rank, world_size)
+
+    def __len__(self):
+        raise NotImplementedError(
+            f"len() is not supported for {self.__class__.__name__}. Please set `trainer.max_steps` to explicitly set the number of steps to train for."
+        )
+
+    def __iter__(self):
+        item = next(iter(self.dataset))
```

Comment:
This is creating a new iter every time, is there an example where this is done elsewhere in NeMo ?

Reply:
Fixed